### PR TITLE
Add Azure AD package for remote write

### PR DESCRIPTION
**Pull request by rakshith210**
This PR contains changes related to Azure AD configuration and Azure AD roundtripper in regard to the following issue:
https://github.com/prometheus/prometheus/issues/11733

Signed-off-by: Rakshith Padmanabha <rapadman@microsoft.com>

### CODE REVIEW

File: storage/remote/azuread/azuread_config.go
Comment (LeviHarrison) [inline storage/remote/azuread/azuread_config.go:26]: nit:
```suggestion
	ClientID string `yaml:"client_id,omitempty"`
```

File: storage/remote/azuread/azuread_roundtripper.go
Comment (LeviHarrison) [inline storage/remote/azuread/azuread_roundtripper.go:22]: general nit:
```suggestion
type azureADRoundTripper struct {
```

File: storage/remote/azuread/constants.go
Comment (LeviHarrison) [inline storage/remote/azuread/constants.go:26]: This constants can probably just be moved into `token_provider.go`, would make things a bit easier to read.

File: storage/remote/azuread/credential.go
Comment (LeviHarrison) [inline storage/remote/azuread/credential.go:23]: As this package is effectively internal to Prometheus, and the rest of Prometheus would have no reason to call this method directly, I think it should be private.

File: storage/remote/azuread/credential_mock.go
Comment (LeviHarrison) [inline storage/remote/azuread/credential_mock.go:30]: Same thing with the private methods/types here.

File: storage/remote/azuread/managed_identity_credential.go
Comment (LeviHarrison) [inline storage/remote/azuread/managed_identity_credential.go:24]: Same thing with the private method here.

File: storage/remote/azuread/token_provider.go
Comment (LeviHarrison) [inline storage/remote/azuread/token_provider.go:46]: Same thing with the private fields/methods here.

**Review by LeviHarrison — COMMENTED** [review]
I left a few small comments, but I my one main question so far is can we add OAuth2 as well, and would that be useful? I know for the Azure AD we have both Managed Identity and OAuth.

Also, if you don't mind, could you hook this up to remote write and just make this PR the whole remote write support. I think it would be easier for testing and it shouldn't be much more on top of this (but let me know if it is and then it might make sense to keep it separate).

**Comment by cstyan** [comment]
should this live in prometheus/common, the same as the aws sigv4 auth code? https://github.com/prometheus/common/tree/main/sigv4

**Comment by roidelapluie** [comment]
The AWS code is in common because it is also used in alertmanager. This code won't be used in alertmanager to the best of my knowledge

**Comment by rakshith210** [comment]
> I left a few small comments, but I my one main question so far is can we add OAuth2 as well, and would that be useful? I know for the Azure AD we have both Managed Identity and OAuth.
> 
> Also, if you don't mind, could you hook this up to remote write and just make this PR the whole remote write support. I think it would be easier for testing and it shouldn't be much more on top of this (but let me know if it is and then it might make sense to keep it separate).

Yes. I do plan to add Azure AD OAuth flow as well. I was thinking of adding it as part of the next PR after this PR is checked in. I wanted to keep this PR concise to just one Auth flow. 

I have changes(fairly minor compared to this change) ready to hook it up to remote write. However, the reason I wanted to take this change too for another PR is because I couldn't directly refer `azuread` package (`github.com/prometheus/prometheus/storage/remote/azuread`) as it is not checked in yet. Let me know if this is fine by you ?

File: storage/remote/azuread/token_provider.go
Comment (rakshith210) [reply storage/remote/azuread/token_provider.go]: made them private

File: storage/remote/azuread/managed_identity_credential.go
Comment (rakshith210) [reply storage/remote/azuread/managed_identity_credential.go]: made them private

File: storage/remote/azuread/credential_mock.go
Comment (rakshith210) [reply storage/remote/azuread/credential_mock.go]: Made the `mockCredential` struct private but `GetToken()` method couldn't be made private  because we are mock the `TokenCredential` interface. https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azcore@v1.3.1/internal/exported#TokenCredential

File: storage/remote/azuread/credential.go
Comment (rakshith210) [reply storage/remote/azuread/credential.go]: Make sense. Made them private

File: storage/remote/azuread/azuread_roundtripper.go
Comment (rakshith210) [reply storage/remote/azuread/azuread_roundtripper.go]: done

File: storage/remote/azuread/azuread_config.go
Comment (rakshith210) [reply storage/remote/azuread/azuread_config.go]: done

File: storage/remote/azuread/go.mod
Comment (rakshith210) [reply storage/remote/azuread/go.mod]: Removed `go.sum` and `go.mod`

File: storage/remote/azuread/README.md
Comment (rakshith210) [reply storage/remote/azuread/README.md]: Yes that's correct, my bad. Removed this line.

File: storage/remote/azuread/Makefile
Comment (rakshith210) [reply storage/remote/azuread/Makefile]: done

File: storage/remote/azuread/azuread_config.go
Comment (csmarchbanks) [inline storage/remote/azuread/azuread_config.go:32]: It would be nice if these comments (and in other files) followed the docstring format, e.g. `Validate validates the config values provided.` https://tip.golang.org/doc/comment. Specifically that they should start with the function name and be complete sentences with a full stop.

File: storage/remote/azuread/azuread_roundtripper.go
Comment (csmarchbanks) [inline storage/remote/azuread/azuread_roundtripper.go:50]: Authorization is spelled incorrectly here.

File: storage/remote/azuread/token_provider.go
Comment (csmarchbanks) [inline storage/remote/azuread/token_provider.go:103]: Since this done in the background do we need locking inside the setters and getters?

File: config/config.go
Comment (csmarchbanks) [inline config/config.go:831]: Just an idea: I wonder if it is worth counting number of auth options enabled rather than multiple if blocks? That might be easier to understand but I am not sure.

File: storage/remote/azuread/azuread_roundtripper.go
Comment (csmarchbanks) [inline storage/remote/azuread/azuread_roundtripper.go:28]: This should be `NewAzureADRoundTripper`.

**Review by csmarchbanks — COMMENTED** [review]
A few comments, I agree with all of the lint warnings/errors as well and would love to have those cleaned up.

Generally this looks ok to me though!

File: storage/remote/azuread/token_provider.go
Comment (LeviHarrison) [inline storage/remote/azuread/token_provider.go:75]: Instead of having a goroutine do this in the background, could we just check if the token has expired on each request, and if so, update it?

**Review by LeviHarrison — COMMENTED** [review]
Looking a lot better!

File: storage/remote/azuread/azuread_config.go
Comment (rakshith210) [reply storage/remote/azuread/azuread_config.go]: Make sense. Updated all the comments to follow the style you suggested.

File: storage/remote/azuread/azuread_roundtripper.go
Comment (rakshith210) [reply storage/remote/azuread/azuread_roundtripper.go]: fixed it

File: storage/remote/azuread/token_provider.go
Comment (rakshith210) [reply storage/remote/azuread/token_provider.go:103]: We are already using `sync/atomic`, which has locking in place and handles the race condition,  to update the values of the accessToken in the `tokenProvider` struct.

```
// setToken handles storing of accessToken value in tokenProvider.
func (tokenProvider *tokenProvider) setToken(token string) {
	var V atomic.Value
	V.Store(token)
	tokenProvider.token = V.Load().(string)
}
```

File: storage/remote/azuread/azuread_roundtripper.go
Comment (rakshith210) [reply storage/remote/azuread/azuread_roundtripper.go]: done

File: config/config.go
Comment (rakshith210) [reply config/config.go:831]: I thought about this, however, it didn't seem pretty straightforward and felt might get more confusing going forward if some new auth was added.

File: storage/remote/azuread/azuread_config.go
Comment (rakshith210) [reply storage/remote/azuread/azuread_config.go]: done

File: storage/remote/azuread/azuread_config.go
Comment (rakshith210) [reply storage/remote/azuread/azuread_config.go]: done

File: storage/remote/azuread/azuread_config.go
Comment (rakshith210) [reply storage/remote/azuread/azuread_config.go]: done

File: storage/remote/azuread/azuread_config.go
Comment (rakshith210) [reply storage/remote/azuread/azuread_config.go]: done

File: storage/remote/azuread/README.md
Comment (rakshith210) [reply storage/remote/azuread/README.md]: Updated the readme with your suggestion

File: config/config.go
Comment (rakshith210) [reply config/config.go]: done

File: storage/remote/azuread/credential.go
Comment (rakshith210) [reply storage/remote/azuread/credential.go]: done

File: storage/remote/azuread/credential.go
Comment (rakshith210) [reply storage/remote/azuread/credential.go]: removed redundant check

File: storage/remote/azuread/managed_identity_credential.go
Comment (rakshith210) [reply storage/remote/azuread/managed_identity_credential.go]: Same here. Removed it

File: storage/remote/azuread/token_provider.go
Comment (rakshith210) [reply storage/remote/azuread/token_provider.go:75]: The idea behind having a goroutine in the background was that we could completely isolate the consumer of the `accessToken` from the refresh logic. Also based to volume of requests we expect in remote write operations, the overhead having background goroutine vs checking for each request seems comparable. Let me know if you would still like me to change the code to check for every request ?

File: storage/remote/azuread/token_provider.go
Comment (LeviHarrison) [inline storage/remote/azuread/token_provider.go:43]: nit: all comments should end with a period (there are a few others that should also be fixed)

```suggestion
	// mu guards access to token.
```

File: storage/remote/azuread/token_provider.go
Comment (LeviHarrison) [inline storage/remote/azuread/token_provider.go:122]: I think it would be more logical to have this check take place after the token is fetched in `getToken()`, instead of both here and also in `valid()`

**Review by LeviHarrison — COMMENTED** [review]
Looking really good now, only a few more comments (also remember to sign of DCO so tests will run).

One general thing: I know I've asked for smaller file consolidations previously, but I think we want to take all the files in the `azuread` package and combine them into just `azuread.go` and `azuread_test.go`. While separate files are helpful for larger, distinct parts of code, I think in this instance the functionally of the package is both small and related enough that it would be simpler to have everything in one file (well two but you know what I mean).

File: storage/remote/azuread/token_provider.go
Comment (rakshith210) [reply storage/remote/azuread/token_provider.go]: fixed the comments in different places.

File: storage/remote/azuread/token_provider.go
Comment (rakshith210) [reply storage/remote/azuread/token_provider.go]: done

**Comment by LeviHarrison** [comment]
Almost done. I went ahead and tested it out, and here are a few observations:
- Could the `cloud` field just default to `AzurePublic`, with the option of configuring it for different clouds if necessary? I feel like the vast majority of users will be using the public Azure cloud so there's no need to trouble them with an additional field.
- For some reason, the client blocks (I think) when requests are made, stopping any blocks from being sent at all. I don't have time right need to dig through the pprof/debugger to find the root cause, but I will at some point if you haven't already gotten to it.

In addition to those, some things that still need to be worked on in the code:
- Configuration documentation needs to be added in the `configuration.md` file.
- I don't think you responded to my [refresh duration comment](https://github.com/prometheus/prometheus/pull/11944#discussion_r1143946492) or changed anything, so I'm going to assume you're still looking into that.

File: docs/configuration/configuration.md
Comment (LeviHarrison) [inline docs/configuration/configuration.md:3306]:
```suggestion
# Cannot be used at the same time as basic_auth, authorization, oauth2, or sigv4.
```

File: docs/configuration/configuration.md
Comment (LeviHarrison) [inline docs/configuration/configuration.md:3309]: We specify defaults using the following syntax, not in the documentation comment:

```suggestion
  # The Azure Cloud.
  [ cloud: <string> | default = AzurePublic ]
```

File: docs/configuration/configuration.md
Comment (LeviHarrison) [inline docs/configuration/configuration.md:3308]: Could you provide the options for clouds here?

File: docs/configuration/configuration.md
Comment (LeviHarrison) [inline docs/configuration/configuration.md:3314]: Only one blank line is needed

```suggestion
```

File: storage/remote/azuread/azuread_config.go
Comment (xy223) [reply storage/remote/azuread/azuread_config.go]: @rakshith210 , will we consider supporting users setting the endpoint url for retrieving aad tokens from the client_id?
We have use cases where instead of using the imds url, we have our own endpoint url for fetching the aad tokens, it'll be really helpful if you could add the endpoint_url, or hostname to the AzureAdConfig.

File: docs/configuration/configuration.md
Comment (rakshith210) [reply docs/configuration/configuration.md]: done

File: docs/configuration/configuration.md
Comment (rakshith210) [reply docs/configuration/configuration.md]: Added cloud options

File: docs/configuration/configuration.md
Comment (rakshith210) [reply docs/configuration/configuration.md]: done

File: docs/configuration/configuration.md
Comment (rakshith210) [reply docs/configuration/configuration.md]: done

File: storage/remote/azuread/azuread_config.go
Comment (rakshith210) [reply storage/remote/azuread/azuread_config.go]: Yes I do plan to add AAD based auth in the next iteration. In this PR, we wanted to have changes limited to just one auth (managed identity in this case).

**Comment by Nexucis** [comment]
I guess the merge / rebase make a bit mess.

@rakshith210 can you try again. That doesn't look right all these changes are related to your PR :)

**Comment by rakshith210** [comment]
> I guess the merge / rebase make a bit mess.
> 
> @rakshith210 can you try again. That doesn't look right all these changes are related to your PR :)

I just rebased my branch with main. It seems to have fixed it

**Comment by Nexucis** [comment]
Perfect thank you !

**Review by csmarchbanks — COMMENTED** [review]
Generally the code looks reasonable to me.

I have been trying to test it and running into some issues though, I don't have great Azure experience or access though so might just be doing something wrong.

One thing I notice is that if I have a misconfiguration in Azure auth it looks like Prometheus is not starting at all. Is that expected/desired? I would expect Prometheus to still start up and just writes to fail.

Example of a failed startup for me:
```
ts=2023-05-16T14:44:08.317Z caller=main.go:534 level=info msg="No time or size retention was set so using the default time retention" duration=15d
ts=2023-05-16T14:44:08.318Z caller=main.go:578 level=info msg="Starting Prometheus Server" mode=server version="(version=2.43.0, branch=users/rakshith210/azuread, revision=418e7b399c662af11001c492f2bf447f24cf65aa)"
ts=2023-05-16T14:44:08.319Z caller=main.go:583 level=info build_context="(go=go1.20.4, platform=linux/amd64, user=root@b11ccdb4fa5d, date=20230516-14:02:33, tags=netgo,builtinassets)"
ts=2023-05-16T14:44:08.319Z caller=main.go:584 level=info host_details="(Linux 5.15.0-1037-azure #44~20.04.1-Ubuntu SMP Mon Apr 24 21:52:51 UTC 2023 x86_64 prom (none))"
ts=2023-05-16T14:44:08.320Z caller=main.go:585 level=info fd_limits="(soft=1048576, hard=1048576)"
ts=2023-05-16T14:44:08.320Z caller=main.go:586 level=info vm_limits="(soft=unlimited, hard=unlimited)"
ts=2023-05-16T14:44:08.324Z caller=web.go:562 level=info component=web msg="Start listening for connections" address=0.0.0.0:9090
ts=2023-05-16T14:44:08.325Z caller=main.go:1019 level=info msg="Starting TSDB ..."
ts=2023-05-16T14:44:08.328Z caller=head.go:595 level=info component=tsdb msg="Replaying on-disk memory mappable chunks if any"
ts=2023-05-16T14:44:08.329Z caller=head.go:676 level=info component=tsdb msg="On-disk memory mappable chunks replay completed" duration=3.3µs
ts=2023-05-16T14:44:08.329Z caller=head.go:684 level=info component=tsdb msg="Replaying WAL, this may take a while"
ts=2023-05-16T14:44:08.332Z caller=tls_config.go:232 level=info component=web msg="Listening on" address=[::]:9090
ts=2023-05-16T14:44:08.332Z caller=tls_config.go:235 level=info component=web msg="TLS is disabled." http2=false address=[::]:9090
ts=2023-05-16T14:44:08.334Z caller=head.go:755 level=info component=tsdb msg="WAL segment loaded" segment=0 maxSegment=1
ts=2023-05-16T14:44:08.335Z caller=head.go:755 level=info component=tsdb msg="WAL segment loaded" segment=1 maxSegment=1
ts=2023-05-16T14:44:08.335Z caller=head.go:792 level=info component=tsdb msg="WAL replay completed" checkpoint_replay_duration=40.001µs wal_replay_duration=5.807404ms wbl_replay_duration=100ns total_replay_duration=6.766721ms
ts=2023-05-16T14:44:08.338Z caller=main.go:1040 level=info fs_type=EXT4_SUPER_MAGIC
ts=2023-05-16T14:44:08.338Z caller=main.go:1043 level=info msg="TSDB started"
ts=2023-05-16T14:44:08.339Z caller=main.go:1224 level=info msg="Loading configuration file" filename=prometheus.yml
ts=2023-05-16T14:44:08.348Z caller=main.go:1250 level=error msg="Failed to apply configuration" err="Failed to get access token: ManagedIdentityCredential authentication failed\nGET http://169.254.169.254/metadata/identity/oauth2/token\n--------------------------------------------------------------------------------\nRESPONSE 400 Bad Request\n--------------------------------------------------------------------------------\n{\n  \"error\": \"invalid_request\",\n  \"error_description\": \"Identity not found\"\n}\n--------------------------------------------------------------------------------\nTo troubleshoot, visit https://aka.ms/azsdk/go/identity/troubleshoot#managed-id"
ts=2023-05-16T14:44:08.424Z caller=main.go:878 level=info msg="Stopping scrape discovery manager..."
ts=2023-05-16T14:44:08.424Z caller=main.go:892 level=info msg="Stopping notify discovery manager..."
ts=2023-05-16T14:44:08.424Z caller=manager.go:1009 level=info component="rule manager" msg="Stopping rule manager..."
ts=2023-05-16T14:44:08.424Z caller=manager.go:1019 level=info component="rule manager" msg="Rule manager stopped"
ts=2023-05-16T14:44:08.424Z caller=main.go:929 level=info msg="Stopping scrape manager..."
ts=2023-05-16T14:44:08.424Z caller=main.go:888 level=info msg="Notify discovery manager stopped"
ts=2023-05-16T14:44:08.424Z caller=main.go:874 level=info msg="Scrape discovery manager stopped"
ts=2023-05-16T14:44:08.425Z caller=main.go:921 level=info msg="Scrape manager stopped"
ts=2023-05-16T14:44:08.425Z caller=manager.go:995 level=info component="rule manager" msg="Starting rule manager..."
ts=2023-05-16T14:44:08.430Z caller=notifier.go:606 level=info component=notifier msg="Stopping notification manager..."
ts=2023-05-16T14:44:08.430Z caller=main.go:1150 level=info msg="Notifier manager stopped"
ts=2023-05-16T14:44:08.430Z caller=main.go:1159 level=error err="error loading config from \"prometheus.yml\": one or more errors occurred while applying the new configuration (--config.file=\"prometheus.yml\")"
```

**Comment by roidelapluie** [comment]
Prometheus should start indeed in this case . We should not make API calls
while reloading the config

Le mar. 16 mai 2023, 16:51, Chris Marchbanks ***@***.***> a
écrit :

**Comment by rakshith210** [comment]
I updated the config structure for azure remote write as follows:
azuread:
  managed_identity:
    client_id: <managed-identity-client-id>

I also fixed the initialization issue faced when wrong config is passed.

**Review by csmarchbanks — COMMENTED** [review]
Made some progress testing it, ran into one issue, otherwise I think this looks good!

File: docs/configuration/configuration.md
Comment (csmarchbanks) [inline docs/configuration/configuration.md:3312]: Also, this configuration needs to be updated with the new nested structure.

File: storage/remote/azuread/azuread.go
Comment (rakshith210) [reply storage/remote/azuread/azuread.go]: Updated this! Now passing context from the request

File: docs/configuration/configuration.md
Comment (rakshith210) [reply docs/configuration/configuration.md]: done

**Review by csmarchbanks — COMMENTED** [review]
It looks like there are some lint failures, otherwise :+1: from me!

**Review by csmarchbanks — APPROVED** [review]
:+1: From me, I will wait until next week to merge in case anyone else has comments.

File: storage/remote/azuread/README.md
Comment (cstyan) [inline storage/remote/azuread/README.md:5]: nit: should we use some formatting if this is a markdown file?

File: storage/remote/azuread/azuread.go
Comment (cstyan) [inline storage/remote/azuread/azuread.go:40]: is the list of which of these exist online somewhere? will we have to add more in the future?

**Review by cstyan — COMMENTED** [review]
some minor nit comments, I trust Chris to make the decision on merging here :+1:

File: storage/remote/azuread/README.md
Comment (rakshith210) [reply storage/remote/azuread/README.md:8]: Added some basic formatting

File: storage/remote/azuread/azuread.go
Comment (rakshith210) [reply storage/remote/azuread/azuread.go:40]: Here is the product catalog where the backend metric store(Managed Prometheus) exists:
https://azure.microsoft.com/en-us/explore/global-infrastructure/products-by-region/?products=monitor

However, these are audiences are per cloud which currently is only the three mentioned above. We will need to add if metrics datastore is made available in new clouds.

**Pull request merged**
